src/Cryptol/REPL/Command.hs

```diff
@@ -391,32 +399,107 @@ dumpTestsCmd outFile str pos fnm =
+  { sample :: Sample -- the sample used for this bin
+  , binIndex :: Integer -- the index of this bin among all bins
+  , bins :: Integer -- the number of total bins
+  , binSize :: Integer -- the number of tests for this bin
+  }
+
 
 -- | Randomly test a property, or exhaustively check it if the number
 -- of values in the type under test is smaller than the @tests@
 -- environment variable, or we specify exhaustive testing.
 qcCmd :: QCMode -> String -> (Int,Int) -> Maybe FilePath -> REPL ()
-qcCmd qcMode "" _pos _fnm =
-  do (xs,disp) <- getPropertyNames
-     let nameStr x = show (fixNameDisp disp (pp x))
-     if null xs
-        then rPutStrLn "There are no properties in scope."
-        else forM_ xs $ \(x,d) ->
-               do let str = nameStr x
-                  rPutStr $ "property " ++ str ++ " "
-                  let texpr = T.EVar x
-                  let schema = M.ifDeclSig d
-                  nd <- M.mctxNameDisp <$> getFocusedEnv
-                  let doc = fixNameDisp nd (pp texpr)
-                  void (qcExpr qcMode doc texpr schema)
-
-qcCmd qcMode str pos fnm =
-  do expr <- replParseExpr str pos fnm
-     (_,texpr,schema) <- replCheckExpr expr
-     nd <- M.mctxNameDisp <$> getFocusedEnv
-     let doc = fixNameDisp nd (ppPrec 3 expr) -- function application has precedence 3
-     void (qcExpr qcMode doc texpr schema)
-
+qcCmd qcMode str pos fnm = do
+  -- (qcMde, doc, texpr, schema) <- case str of 
+  case str of
+    "" -> do
+      (xs,disp) <- getPropertyNames
+      let nameStr x = show (fixNameDisp disp (pp x))
+      if null xs
+      then do
+        rPutStrLn "There are no properties in scope."
+      else
+        forM_ xs $ \(x,d) -> do
+          let str = nameStr x
+          rPutStr $ "property " ++ str ++ " "
+          let texpr = T.EVar x
+          let schema = M.ifDeclSig d
+          nd <- M.mctxNameDisp <$> getFocusedEnv
+          let doc = fixNameDisp nd (pp texpr)
+          go doc Nothing texpr schema
+    _ -> do
+      expr <- replParseExpr str pos fnm
+      (_,texpr,schema) <- replCheckExpr expr
+      nd <- M.mctxNameDisp <$> getFocusedEnv
+      let doc = fixNameDisp nd (ppPrec 3 expr) -- function application has precedence 3
+      go doc (Just expr) texpr schema
+  where
+    go :: Doc -> Maybe (P.Expr P.PName) -> T.Expr -> T.Schema -> REPL ()
+    go doc mb_expr texpr schema = case qcMode of
+      QCExhaust -> do
+        testNum <- (toInteger :: Int -> Integer) <$> getKnownUser "tests"
+        (val, ty) <- replEvalCheckedExpr' texpr schema
+        void (qcExpr qcMode doc texpr schema testNum val ty Nothing)
+      QCRandom -> do
+        testNum <- (toInteger :: Int -> Integer) <$> getKnownUser "tests"
+        useLitSampling <- getKnownUser "literalSampling" :: REPL Bool
+        if  useLitSampling &&
+            not (null (T.sVars schema)) &&
+            all ((T.KNum ==) . T.tpKind) (T.sVars schema)
+          then do
+          case mb_expr of
+            Nothing -> do
```

Comment:
Not sure I understand this case. Why do we only do the sampling in the `Just` case?